### PR TITLE
pre-alloc optimized xw product output vector

### PR DESCRIPTION
**Pull request by havardpe**
@arnej27959 please review
@lesters FYI

### CODE REVIEW

File: eval/src/vespa/eval/tensor/dense/dense_xw_product_function.cpp [hunk 1]
```diff
@@ -85,17 +85,24 @@ DenseXWProductFunction::DenseXWProductFunction(const eval::ValueType &resultType
                                                size_t resultSize,
                                                bool matrixHasCommonDimensionInnermost)
     : eval::tensor_function::Op2(resultType, vector_in, matrix_in),
-      _self(resultType, vectorSize, resultSize, matrixHasCommonDimensionInnermost)
+      _vectorSize(vectorSize),
+      _resultSize(resultSize),
+      _commonDimensionInnermost(matrixHasCommonDimensionInnermost)
 {}
 
 namespace {
```
Comment (arnej27959) [inline eval/src/vespa/eval/tensor/dense/dense_xw_product_function.cpp]: remove leftover empty namespace

File: eval/src/vespa/eval/tensor/dense/dense_xw_product_function.cpp [hunk 2]
```diff
@@ -85,17 +85,24 @@ DenseXWProductFunction::DenseXWProductFunction(const eval::ValueType &resultType
...
 } // namespace <unnamed>
 
 eval::InterpretedFunction::Instruction
-DenseXWProductFunction::compile_self(Stash &) const
+DenseXWProductFunction::compile_self(Stash &stash) const
 {
-    return eval::InterpretedFunction::Instruction(my_op, (uint64_t)(&_self));
+    Self &self = stash.create<Self>(result_type(), _vectorSize, _resultSize, stash);
+    if (_commonDimensionInnermost) {
+        return eval::InterpretedFunction::Instruction(my_op<true>, (uint64_t)(&self));
```
Comment (arnej27959) [inline eval/src/vespa/eval/tensor/dense/dense_xw_product_function.cpp]: we could also write this as
```
    auto op = _commonDimensionInnermost ? my_op<true> : my_op<false>;
    return eval::InterpretedFunction::Instruction(op, (uint64_t)(&self));
```

**Review by arnej27959 — APPROVED** [review]
minor suggestions only

**Comment by havardpe** [comment]
please have another look

**Pull request merged**